contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -927,6 +928,263 @@ checksum(PG_FUNCTION_ARGS)
         PG_RETURN_INT32(result);
 }
 
+/*
+ * object_id
+ * 	Returns the object ID with object name and object type as input where object type is optional
+ * Returns NULL
+ * 	if input is NULL
+ * 	if there is no such object
+ * 	if user don't have right permission
+ * 	if any error occured
+ */
+Datum
+object_id(PG_FUNCTION_ARGS)
+{	
+	char			*db_name, *schema_name, *object_name;
+	char			*physical_schema_name;
+	char			*input;
+	char			*object_type = NULL;
+	char			**splited_object_name;
+	Oid			schema_oid;
+	Oid			user_id = GetUserId();
+	Oid			result = InvalidOid;
+	bool			is_temp_object;
+	int			i;
+
+	if(PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+	else 
+		input = text_to_cstring(PG_GETARG_TEXT_P(0));
+	if(!PG_ARGISNULL(1))
+	{
+		object_type = text_to_cstring(PG_GETARG_TEXT_P(1));
+		i = strlen(object_type);
+		if (i > 2)
+		{
+			pfree(input);
+			pfree(object_type);
+			PG_RETURN_NULL();
+		}
+		else if (i == 2 && isspace((unsigned char) object_type[1]))
+		{
+			object_type[1] = '\0';
+		}
+	}
+	/* strip trailing whitespace from input */
+	i = strlen(input);
+	while (i > 0 && isspace((unsigned char) input[i - 1]))
+		input[--i] = '\0';
+	
+	/* length should be restricted to 4000 */
+	if (i > 4000)
+		ereport(ERROR,
+			(errcode(ERRCODE_STRING_DATA_RIGHT_TRUNCATION),
+			errmsg("String or binary data would be truncated.")));
+
+	/* resolve the three part name */
+	splited_object_name = split_object_name(input);
+	db_name = splited_object_name[1];
+	schema_name = splited_object_name[2];
+	object_name = splited_object_name[3];
+
+	/* downcase identifier if needed */
+	if (pltsql_case_insensitive_identifiers)
+	{	
+		db_name = downcase_identifier(db_name, strlen(db_name), false, false);
+		schema_name = downcase_identifier(schema_name, strlen(schema_name), false, false);
+		object_name = downcase_identifier(object_name, strlen(object_name), false, false);
+	}
+	
+	for (int i = 0; i < 4; i++)
+		pfree(splited_object_name[i]);
```

Comment:
this is incorrect if pltsql_case_insensitive_identifiers is false, we are freeing memory still referred by *_name objects

Reply:
Thanks for pointing it.